test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,424 @@
+        finally:
+            # Clear mempool
+            self.generate(self.nodes[0], 1)
+            self.nodes[0].disconnect_p2ps()
+            self.mocktime += LONG_TIME_SKIP
+            self.nodes[0].setmocktime(self.mocktime)
+    return wrapper
+
+class PeerTxRelayer(P2PTxInvStore):
+    """A P2PTxInvStore that also remembers all of the getdata and tx messages it receives."""
+    def __init__(self):
+        super().__init__()
+        self._tx_received = []
+        self._getdata_received = []
+
+    @property
+    def tx_received(self):
+        with p2p_lock:
+            return self._tx_received
+
+    @property
+    def getdata_received(self):
+        with p2p_lock:
+            return self._getdata_received
+
+    def on_tx(self, message):
+        self._tx_received.append(message)
+
+    def on_getdata(self, message):
+        self._getdata_received.append(message)
+
+    def wait_for_parent_requests(self, txids):
+        """Wait for requests for missing parents by txid with witness data (MSG_WITNESS_TX or
+        WitnessTx). Requires that the getdata message match these txids exactly; all txids must be
+        requested and no additional requests are allowed."""
+        def test_function():
+            last_getdata = self.last_message.get('getdata')
+            if not last_getdata:
+                return False
+            return len(last_getdata.inv) == len(txids) and all([item.type == MSG_WITNESS_TX and item.hash in txids for item in last_getdata.inv])
+        self.wait_until(test_function, timeout=10)
+
+    def assert_no_immediate_response(self, message):
+        """Check that the node does not immediately respond to this message with any of getdata,
+        inv, tx. The node may respond later.
+        """
+        prev_lastmessage = self.last_message
+        self.send_and_ping(message)
+        after_lastmessage = self.last_message
+        for msgtype in ["getdata", "inv", "tx"]:
+            if msgtype not in prev_lastmessage:
+                assert msgtype not in after_lastmessage
+            else:
+                assert_equal(prev_lastmessage[msgtype], after_lastmessage[msgtype])
+
+    def assert_never_requested(self, txhash):
+        """Check that the node has never sent us a getdata for this hash (int type)"""
+        for getdata in self.getdata_received:
+            for request in getdata.inv:
+                assert request.hash != txhash
+
+class OrphanHandlingTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [[]]
+        self.mocktime = int(time.time())
+
+    def create_parent_and_child(self):
+        """Create package with 1 parent and 1 child, normal fees (no cpfp)."""
+        parent = self.wallet.create_self_transfer()
+        child = self.wallet.create_self_transfer(utxo_to_spend=parent['new_utxo'])
+        return child["tx"].getwtxid(), child["tx"], parent["tx"]
+
+    def fastforward(self, seconds):
```

Comment:
Maybe name this something that's a bit more clearly related to mocktime (`bumpmocktime`?) and add it to `BitcoinTestFramework` ?

Reply:
Good point, added a `bumpmocktime` to test framework, I'll probably use it in other tests in the future.